frontend/javascripts/oxalis/model/bucket_data_handling/data_cube.js

```diff
@@ -367,6 +373,143 @@ class DataCube {
+      if (
+        !areBoundingBoxesOverlappingOrTouching(currentBucket.getBoundingBox(), viewportBoundings)
+      ) {
+        continue;
+      }
+      const bucketData = currentBucket.getOrCreateData();
+      const initialVoxelIndex = this.getVoxelIndex(initialVoxelInBucket, zoomStep);
+      if (bucketData[initialVoxelIndex] !== sourceCellId) {
+        // Ignoring neighbour buckets whose cellId at the initial voxel does not match the source cell id.
+        continue;
+      }
+      // Add the bucket to the current volume undo batch, if it isn't already part of it.
+      currentBucket.markAndAddBucketForUndo();
+      // Mark the initial voxel.
+      bucketData[initialVoxelIndex] = cellId;
+      // Use a VoxelNeighborStack2D to iterate over the bucket in 2d and using bucket-local addresses and not global addresses.
+      const neighbourVoxelStack = new VoxelNeighborStack2D(get2DAddress(initialVoxelInBucket));
+      // Iterating over all neighbours from the initialAddress.
+      while (!neighbourVoxelStack.isEmpty()) {
+        const neighbours = neighbourVoxelStack.popVoxelAndGetNeighbors();
+        for (let neighbourIndex = 0; neighbourIndex < neighbours.length; ++neighbourIndex) {
+          const neighbourVoxel = neighbours[neighbourIndex];
+          let neighbourVoxel3D = get3DAddress(neighbourVoxel, seedVoxel);
+
+          // If the current neighbour is not in the current bucket, calculate its
+          // bucket's zoomed address and add the bucket to bucketsToFill.
+          // adjustedNeighbourVoxel is a copy of neighbourVoxel whose value are robust
+          // against the modulo operation used in getVoxelOffset.
+          const {
+            neighbourBucketAddress,
+            adjustedVoxel: adjustedNeighbourVoxel,
+          } = this.isVoxelInsideBucket(
+            neighbourVoxel,
+            currentBucket,
+            dimensionsToIterateOver,
+            zoomStep,
+          );
+          neighbourVoxel3D = get3DAddress(adjustedNeighbourVoxel, seedVoxel);
+          if (neighbourBucketAddress) {
+            // Add the bucket to the list of buckets to flood fill.
+            const neighbourBucket = this.getOrCreateBucket(neighbourBucketAddress);
+            if (neighbourBucket.type !== "null") {
+              bucketsToFill.push([
+                neighbourBucket,
+                this.getVoxelOffset(neighbourVoxel3D, zoomStep),
+              ]);
+            }
+          } else {
+            // Label the current neighbour and add it to the neighbourVoxelStack to iterate over its neighbours.
+            const neighbourVoxelIndex = this.getVoxelIndex(neighbourVoxel3D, zoomStep);
+            if (bucketData[neighbourVoxelIndex] === sourceCellId) {
+              bucketData[neighbourVoxelIndex] = cellId;
+              neighbourVoxelStack.pushVoxel(neighbourVoxel);
+            }
+          }
+        }
+      }
+      currentBucket.trigger("bucketLabeled");
```

Comment:
I think, this should also be covered when iterating over `bucketsToAddToPushQueue` later.